CNV-9852: New content to configure a watchdog (#32023)
openshift/openshift-docs: modules/virt-defining-a-watchdog.adoc

// Module included in the following assemblies:
//
// * virt/virtual_machines/advanced_vm_management/virt-configuring-a-watchdog.adoc

[id="virt-defining-a-watchdog"]
= Defining a watchdog device

.Procedure

> . Choose the watchdog action. This defines how the watchdog proceeds when the operating system (OS) no longer responds.

Reviewer: How does the user choose a watchdog action? Select from a list? Select a check box? Please see IBM style guide for the appropriate verb. I can't tell whether this is a UI or not. If it is a UI, it's a good idea to start with "In the OpenShift console, ..." so that the user has a context for this step.

Reviewer: On second thought, it looks like these are options for the user to put in the yaml. So mentally selecting an action is not really a procedure step. I recommend adding them as callouts to the yaml.

+
.Available actions

> [horizontal]

Reviewer: I suggest you put these definitions into the introduction of your module, before the procedure.

`poweroff`:: The virtual machine (VM) powers down immediately.

> `reset`:: The VM reboots in place, and the guest OS cannot react. Because the length of time required for the guest OS to reboot can cause liveness probes to timeout, you should avoid their use.

Reviewer: `The VM reboots in place, and the guest OS cannot react.` - Remove comma.

`Because the length of time required for the guest OS to reboot can cause liveness probes to timeout, you should avoid their use.` 
`timeout` > time out
"avoid" or "limit" their use? If the user needs to avoid this setting, why do we offer it to them? It would be helpful to explain this further, perhaps by telling the user when this should be used.

Author: @kbidarkar @stu-gott 
Can you provide additional info here?

Reviewer 2: Liveness probes failing aren't a catastrophe at this point. The VM was already wedged and the watchdog kicked in. The impact is that from the user's point of view, it can take longer for the VM to be restarted.

How about this?

`reset`:: The VM reboots in place, and the guest OS cannot react. This can cause liveness probes to time out, which would make the VM take even longer to come back online. Use of this option is discouraged for this reason.

Reviewer 2: To be clear, the VM takes longer because cluster-level protections will notice the liveness probe failed and forcibly reschedule it.

Author: I can work with that.

> `shutdown`:: The VM gracefully powers down gracefully by stopping all services.

Reviewer 3: `The VM gracefully powers down by stopping all services.`

Probably a typo, feel we remove the extra `gracefully` and it should be good.

Author: Can do, thanks!

> . Create a YAML file and apply it to the cluster.

Reviewer: Please tell the user how to do this. The user needs to create a yaml file, save it, and then use the 'oc apply' command to update the virtual machine CR.

+

> The example below configures the `i6300esb` watchdog device on a RHEL8 VM with the poweroff action and exposes the device as `/dev/watchdog`.

Reviewer: It would be more efficient to add these examples to the callouts, since you need to call out the user-defined parameter values any way.

Author: Added a callout and moved this below example.

+
[source,yaml]
----
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm2-rhel84-watchdog

>   name: vm2-rhel84-watchdog

Reviewer: `vm2-rhel84-watchdog` to `<vm2-rhel84-watchdog>`

spec:
  running: false
  template:
    metadata:
     labels:
        kubevirt.io/vm: vm2-rhel84-watchdog
    spec:
      domain:
        devices:
          watchdog:

>             name: mywatchdog

Reviewer: Change `name: mywatchdog` to `name: <watchdog>`

            i6300esb:

>               action: "poweroff"

Reviewer: This is where you would put a callout:
action: <action> <1>
...
<1> Specify the watchdog action. Valid options are `poweroff`, etc.

>           disks:

Reviewer: I think you could omit the rest of this yaml. Just put in `...` after the watchdog section.

          - disk:
              bus: virtio
            name: containerdisk
          - disk:
              bus: virtio
            name: cloudinitdisk
        machine:

>           type: ""

Reviewer: Is `""`  for `type` here for a specific reason?

        resources:
          requests:
            memory: "1Gi"
      terminationGracePeriodSeconds: 0
      volumes:
      - containerDisk:
          image: registry.redhat.io/rhel8/rhel-guest-image
        name: containerdisk
      - cloudInitNoCloud:
          userData: |-
            #cloud-config

>             password: redhat

Reviewer: Change `redhat` to `<password>`.
Remove the `#cloud-config` comment unless it is really needed. I doubt it.

            chpasswd: { expire: False }
        name: cloudinitdisk
----

This device can now be used by the watchdog binary.

.Verification

> Confirm watchdog functionality on non-production systems.

Reviewer: You can remove this sentence. Since the heading is 'Verification' and you have a callout saying that the step should only be run on non-production systems, this sentence is redundant.

> [WARNING]

Reviewer: I think we stopped using this callout because IBM says that `warning` should only be used in cases of real danger like "Warning: Disconnect the power before removing the hard drive." You can use IMPORTANT here.

====

> This procedure is provided for testing the functionality of watchdogs only and must not be run on production machines.

Reviewer: Suggest changing `functionality of watchdogs` to `watchdog functionality`.

====

> . Run the following command on the virtual machine to confirm the VM identified the watchdog card:

Reviewer: You have virtual machine/VM twice. I think it could be made more concise.

Suggestion: Verify that the VM has identified the watchdog device:

+
[source,terminal]
----
$ lspci | grep watchdog -i
----

> . Run one of the following commands to confirm the watchdog is active:

Reviewer: I'm not sure I would even make this into a choice. I would just make the kernel panic and the process kill into separate steps. 

But if you want to keep it as a choice, make the two commands into bullet list items.

+
Trigger a kernel panic:
+
[source,terminal]
----
$ echo c > /proc/sysrq-trigger
----
+
Terminate the watchdog service:
+
[source,terminal]
----
$ kill -9 pgrep watchdog
----

.Additional information

> * link:https://docs.openshift.com/container-platform/4.7/applications/application-health.html[Monitoring application health by using health checks]

Reviewer: First, you cannot link to the upstream OCP docs site. That domain is going to expire soon, so all the links will be dead. We only link to docs in Customer Portal.

Second, this would be an xref, not a link, because CNV is part of OCP docs. so it needs to be xref:../applications/application-health.adoc#<anchor>_<context>. You will also have to ensure that you get the path correct because CNV assemblies have a complicated structure. So don't be surprised if your build fails a few times. If you are really stuck, you can ping me.

+
.Available actions
[horizontal]
`poweroff`:: The virtual machine (VM) powers down, and the guest operating system (OS) cannot react.
`reset`:: The VM reboots in place, and the guest OS cannot react. Because the length of time required for the guest OS to reboot can cause liveness probes to timeout, you should avoid their use.
`shutdown`:: The VM powers off gracefully and becomes non-runnable.

. Step 2? Create a YAML file and apply it to the cluster?
. Step 3?

The example below configures the i6300esb watchdog device on a Fedora VM with the poweroff action and exposes the device as `/dev/watchdog`.

[source,yaml]
----
---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-with-watchdog
  name: vmi-with-watchdog
spec:
  domain:
    devices:
      watchdog:
        name: mywatchdog
        i6300esb:
          action: "poweroff"
      disks:
      - disk:
          bus: virtio
        name: containerdisk
    machine:
      type: ""
    resources:
      requests:
        memory: 512M
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: kubevirt/fedora-cloud-container-disk-demo
    name: containerdisk
----

> This device can now be used by the watchdog binary. The following example will send a heartbeat every two seconds to `/dev/watchdog` and execute the defined action (in this case, `poweroff`) after four seconds without a heartbeat. This command must be run inside the guest.

Reviewer 4: Maybe you can rewrite this slighlty. I think the important point is, that this action will be executed when the binary does not call `/dev/watchdog` anymore. It sounds a little like the binary would be responsible for the `poweroff` action.

// Module included in the following assemblies:
//
// * virt/virtual_machines/advanced_vm_management/virt-configuring-a-watchdog.adoc

[id="virt-defining-a-watchdog"]

> = Defining a watchdog

Reviewer 5: ```suggestion
= Defining a watchdog device
```

.Procedure

> . Choose the watchdog action. This defines how the watchdog proceeds when the OS can no longer respond.

Reviewer 5: minor suggestions; "OS" may mean many things.
```suggestion
. Choose the watchdog action. This defines how the watchdog proceeds when the guest operating system no longer responds.
```

+
.Available actions
[horizontal]
`poweroff`:: The virtual machine (VM) powers down, and the guest OS cannot react.

> `reset`:: The VM reboots in place, and the guest OS cannot react. Because the length of time required for the guest OS to reboot can cause liveness probes to timeout, you should avoid their use.

Reviewer 5: "liveness probes" or "liveliness probes"? I do not know myself, but please be consistent.

> `shutdown`:: The VM powers off gracefully and becomes non-runnable.

Reviewer 2: I find this statement confusing. To say a VM has become "non-runnable" implies that I cannot re-start it. Is that the case here?

Author: @stu-gott This language was pulled from another doc I believe re: the available actions. Do you have a suggestion for clarity?

Reviewer 2: The path forward is not clear to me. Further re-reading of the 3 actions might imply with wording is entirely correct and intentional. Do you know who was involved in creating the document you referenced?

Reviewer 2: @rmohr do you have insight into the particular wording for these three actions?

Author: @stu-gott @rmohr Although I know I pulled the "non-runnable" wording from somewhere else, I can't find where. Perhaps it was a fever dream. Happy to remove that and end this at powers off gracefully - just let me know.

Reviewer 2: The thing that has me confused is that I can't understand a clear distinction between poweroff and shutdown.

Author: @stu-gott From my notes it appears the guest OS can react in shutdown, unlike poweroff. This is probably worth calling out for clarity. It also looks like it needs to be unlocked in shutdown but maybe not in poweroff? Rewrite suggestion: `The VM powers off gracefully if it is unlocked. The guest OS can react. `

Reviewer 3: From what I observed during my testing,

a) poweroff instantly brings down the VM

b) Shutdown gracefully brings down the VM, by stopping all the services.

Though in both the cases a) and b) , the VM automatically got into running state when the watchdog process was killed.

This was when in the vm spec, it had spec.running = False

Reviewer 2: ok. Thanks for clarifying @kbidarkar.

@sabrinajess I suggest rewording the line this way:

```
`shutdown`:: The VM powers off gracefully.
```

. Create a YAML file and apply it to the cluster.
+

> The example below configures the i6300esb watchdog device on a Fedora VM with the poweroff action and exposes the device as `/dev/watchdog`.

Reviewer 5: ```suggestion
The example below configures the `i6300esb` watchdog device on a RHEL8 VM with the poweroff action and exposes the device as `/dev/watchdog`.
```
Can we use a supported OS?

+
[source,yaml]
----
---
apiVersion: kubevirt.io/v1

> kind: VirtualMachineInstance

Reviewer 5: I think that giving an example with a VirtualMachine (not a throw-away VMI) would be more helpful for users

metadata:
  labels:
    special: vmi-with-watchdog
  name: vmi-with-watchdog
spec:
  domain:
    devices:
      watchdog:
        name: mywatchdog
        i6300esb:
          action: "poweroff"
      disks:
      - disk:
          bus: virtio
        name: containerdisk
    machine:
      type: ""
    resources:
      requests:
        memory: 512M
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:

>       image: kubevirt/fedora-cloud-container-disk-demo

Reviewer 5: ```suggestion
      image: registry.redhat.io/rhel8/rhel-guest-image
```

>     name: containerdisk

Reviewer 3: This sample yaml file, does not seem to work, it failed with the below error.

]$ oc apply -f vm-rhel8-watchdog-docs.yaml 
Error from server (InternalError): error when creating "vm-rhel8-watchdog-docs.yaml": Internal error occurred: failed calling webhook "mutatevirtualmachines.kubemacpool.io": Post "https://kubemacpool-service.openshift-cnv.svc:443/mutate-virtualmachines?timeout=10s": EOF

I had shared an example yaml file, in the Jira issue of watchdog.

----

This device can now be used by the watchdog binary. The following example will send a heartbeat every two seconds to `/dev/watchdog` and execute the defined action after four seconds without a heartbeat. This command must be run inside the guest.

[source,terminal]
----

> $ watchdog -t 2000ms -T 4000ms /dev/watchdog

Reviewer 5: on my Fedora, this does not work, I believe it is very old. Please drop this example.
```
$ watchdog -t 2000ms -T 4000ms /dev/watchdog
Option -t is no longer valid, please specify it in /etc/watchdog.conf.
```

Reviewer 3: I tested this with RHEL8 containerDisk Image. Left few comments on the Jira issue.
- I was unable to run this command, `watchdog -t 2000ms -T 4000ms /dev/watchdog`
- Probably this command is for another version of watchdog.
- Watchdog shipped with RHEL8 does not seem to support the above options.

----

.Verification
. Run the following command on the virtual machine to confirm the VM identified the watchdog card:

[source,terminal]
----
$ lspci | grep watchdog -i
----

> . Run one of the following commands to confirm the watchdog is active:

Reviewer 3: Was curious to know, how the doc existed for RHV and came across the below:
https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.4/html/virtual_machine_management_guide/sect-configuring_a_watchdog

Feel it would be good to have a warning before this step, similar to what we have for in RHV.

Author: Is this the warning you want to mirror?

WARNING
This procedure is provided for testing the functionality of watchdogs only and must not be run on production machines.

Reviewer 3: Yes.

Trigger a kernel panic:
[source,terminal]
----
$ echo c > /proc/sysrq-trigger
----

Terminate the watchdog service:
[source,terminal]
----
$ kill -9 pgrep watchdog

> ----

Reviewer 5: Please move this to after `systemctl enable` command. The user should verify their actual service.

// Module included in the following assemblies:
//
// * virt/virtual_machines/advanced_vm_management/virt-configuring-a-watchdog.adoc

[id="virt-defining-a-watchdog"]
= Defining a watchdog device

Define how the watchdog proceeds when the operating system (OS) no longer responds.

.Available actions
[horizontal]
`poweroff`:: The virtual machine (VM) powers down immediately.
`reset`:: The VM reboots in place and the guest OS cannot react. Because the length of time required for the guest OS to reboot can cause liveness probes to timeout, use of this option is discouraged. This timeout can extend the time it takes the VM to reboot if cluster-level protections notice the liveness probe failed and forcibly reschedule it.
`shutdown`:: The VM gracefully powers down by stopping all services.

.Procedure

. Create a YAML file with the following contents:

+
[source,yaml]
----

> ---

Reviewer 6: I think you can remove this `---` because most of our YAML file examples start with the `apiVersion` line.
